### DISCUSSION TITLE
Precommit in Kedro template

### DISCUSSION THREAD

**Discussion by lucasjamar** [kedro-org/kedro#1347] [Idea]
Continuing this discussion over from discord.
The idea would be to add precommit to the kedro cookicutter template. What do you guys think? Some of the benefits i see would be:

1. No longer need to keep isort, black, flake8  in requirements.in. Precommit detects the packages in .pre-commit-config.yaml and install thems in a separate env.
https://github.com/kedro-org/kedro/issues/1293
https://discord.com/channels/778216384475693066/779008389027332136/951480920966311996
3. Could be argued that this certain issues related to deployment. Users would no longer need to do linting in their github actions or gitlab ci.
4. Additional best practices could be added: pyupgrade, trailing-whitespace, end-of-file-fixer and check-yaml with little complexity.
5. Would be incentive for beginners to always create a git repo which could be argued is the most important best practice of all.

**Comment by deepyaman** [comment]
Very supportive of the idea of adding `pre-commit` to the template. Kedro has always been opinionated, but it may be worth thinking about adding "advanced" flags like `--no-pre-commit` in case people don't want it.

> Could be argued that this certain issues related to deployment. Users would no longer need to do linting in their github actions or gitlab ci.

Disagree that users no longer need to do linting in CI, as people can always `git commit --no-verify`. Never trust your users to enforce anything that the CI doesn't!

Agree on the rest of the points.

**Reply by lucasjamar** [reply]
Very interesting. I didnt know the `git commit --no-verify` flag. Ill be sure to not share that info with colleagues :P

**Reply by antonymilne** [reply]
Also note that it's very easy to evade pre-commit checks just by making a commit through the Github UI rather than your own dev environment. Not an argument against pre-commit at all, but as per @deepyaman you should not rely on pre-commit - always enforce checks through CI as well.

**Comment by antonymilne** [comment]
Let me just copy my comment from Discord to here for the record 🙂 

This is very similar to @WaylonWalker's suggestions https://github.com/kedro-org/kedro/discussions/844 and https://github.com/kedro-org/kedro-starters/pull/40. I like pre-commit too, but my general concern here that supporting it in the kedro template might become a maintenance burden for us. Given that it's already straightforward to create and use custom starters outside the official kedro list (through `kedro new -s`), I would suggest that is a good way forward is to create an unofficial starter for this that lives outside kedro-org (maybe even building on https://github.com/kedro-org/kedro-starters/pull/40).

Personally I would propose:
- remove `kedro lint` (under discussion in #1293)
- keep the default `kedro new` template very minimal with no `black`, `flake8` etc.
- make a new, *community-maintained* starter which includes all the stuff that people want like pre-commit. It can even ship with custom CLI commands like `kedro lint` if they're still wanted
- you can then use this via `kedro new -s  http://github.com/path/to/starter.git`
- even better, after @Galileo-Galilei's work in https://github.com/kedro-org/kedro/pull/1265 you could make it into a pip-installable package and would be able to `pip install kedro-starter-precommit` and then see it in `kedro starter list` and use it as `kedro new -s precommit`
